feat: add browser extension

svrnm opened the pull request:
## Which problem is this PR solving?
As discussed in #490, this PR provides code for a browser extension, that can be used to inject OpenTelemetry in an arbitrary website. The original repository at https://github.com/svrnm/opentelemetry-browser-extension/ is owned by me.

## Short description of the changes
- Add browser-extension/ at the root level of the repository
- Add the source code to build the extension

obecny commented on browser-extension/src/ui/index.tsx:
```diff
@@ -0,0 +1,361 @@
+            >
+              Exporter Settings
+            </Typography>
+            <Grid container spacing={2}>
```
I would split this into smaller components 
So it could look like this
```
<ExporterOption label="Console" url="https://www.npmjs.com/package/@opentelemetry/tracing" toggle={exporters[ExporterType.CONSOLE]}/>

<ExporterOption label="Zipkin" url="https://www.npmjs.com/package/@opentelemetry/tracing" toggle={exporters[ExporterType.ZIPKIN]}/>

<ButtonSave onClick={this.handleSaveSettings}/>
```
etc.
It will be more readable

obecny commented on browser-extension/src/ui/styles.ts:
```diff
@@ -0,0 +1,41 @@
+      marginLeft: theme.spacing(2),
+      marginRight: theme.spacing(2),
+      [theme.breakpoints.up(400 + theme.spacing(2) * 2)]: {
+        width: 800,
```
what if user browser width is 700 ?, should this depend on window width too ?

svrnm replied:
I copied&pasted that to get started with styling from material-ui (since I am not really good at those things...). Having this makes no sense, I removed it.

obecny commented on browser-extension/src/utils/EventEmitterSpanExporter.ts:
```diff
@@ -0,0 +1,46 @@
+    spans: ReadableSpan[],
+    resultCallback: (result: ExportResult) => void
+  ): void {
+    if (this._stopped) {
```
Nit: 
```suggestion
    if (this._isShutdown) {
```

obecny commented on browser-extension/src/utils/EventEmitterSpanExporter.ts:
```diff
@@ -0,0 +1,46 @@
+  extends EventTarget
+  implements SpanExporter
+{
+  protected _stopped = false;
```
```suggestion
  protected _isShutdown = false;
```

obecny reviewed (commented):
first pass, nice work, I have raised few issues.
1. Please move also the package to folder `packages` for now, if we figure out better place it can be moved later, but in general I would not keep it in main folder.
2. Can you please add some screenshots how it looks so we can see it too ?

svrnm commented:
Thanks for taking the time to look into this!

> 1. Please move also the package to folder `packages` for now, if we figure out better place it can be moved later, but in general I would not keep it in main folder.

Sure!

> 2. Can you please add some screenshots how it looks so we can see it too ?

Yes, I can expand the "Usage" guide with some screenshots: "what to do" and "what to expect"

svrnm commented on browser-extension/package.json:
```diff
@@ -0,0 +1,66 @@
+{
+  "name": "@opentelemetry/browser-extension",
```
I moved it into packages and going to name it "opentelemetry-injection-browser-extension", since the purpose is injecting opentelemetry into arbitrary websites.

svrnm replied:
based on @dyladan's comment I renamed it again to opentelemetry-browser-extension-autoinjection

svrnm commented on browser-extension/webpack.config.ts:
```diff
@@ -0,0 +1,153 @@
+// eslint-disable-next-line node/no-unpublished-import
+import * as HtmlWebpackPlugin from 'html-webpack-plugin';
+
+const targetMV3 = {
```
done.

dyladan commented:
In general we try to name packages `opentelemetry-category-name`. `opentelemetry-injection-browser-extension` is `opentelemetry-name-category`. I would prefer the name `opentelemetry-browser-extension-injection`, `opentelemetry-browser-extension-autoinjection`, `opentelemetry-browser-ext-injection` or similar.

svrnm commented:
> In general we try to name packages `opentelemetry-category-name`. `opentelemetry-injection-browser-extension` is `opentelemetry-name-category`. I would prefer the name `opentelemetry-browser-extension-injection`, `opentelemetry-browser-extension-autoinjection`, `opentelemetry-browser-ext-injection` or similar.

Ah okay, that makes sense! I'll change that.

mhennoch commented:
I'll try to take a look tomorrow, we actually have something similar but haven't gotten around upstreaming it.

svrnm commented on browser-extension/webpack.config.ts:
```diff
@@ -0,0 +1,153 @@
+
+const targetMV3 = {
+  entry: {
+    ui: './src/ui/index.tsx',
```
fixed

svrnm commented on browser-extension/package.json:
```diff
@@ -0,0 +1,66 @@
+  "version": "0.1.1",
+  "description": "A browser extension that injects opentelemetry tracers into any website",
+  "scripts": {
+    "compile": "npx webpack --mode=production",
```
I added "clean" and "prepare", not sure about version:update. Since I also extract the package name from package.json at compile time, I have the following right now in `ui/index.tsx`:

```javascript
...
const packageJson = require('../../package.json');
...
              <Typography variant="h6" color="inherit" noWrap>
                {capitalCase(packageJson.name)} ({packageJson.version})
              </Typography>
...
``` 

I could make it consistent with the other packages and work with a version.ts, I don't have a real preference.

svrnm commented on browser-extension/package.json:
```diff
@@ -0,0 +1,66 @@
+{
+  "name": "@opentelemetry/browser-extension",
+  "version": "0.1.1",
```
fixed.

svrnm commented on browser-extension/src/background/index.ts:
```diff
@@ -0,0 +1,84 @@
+      request.extensionId === chrome.runtime.id &&
+      sender.id === chrome.runtime.id
+    ) {
+      const spans = JSON.parse(request.spans);
```
good point. For now I added a try catch block (not really making any difference..), this code is still not working the way I want it to be so it's currently not used. I might need to add some error display later on.

svrnm commented on browser-extension/src/background/index.ts:
```diff
@@ -0,0 +1,84 @@
+      }
+
+      chrome.tabs.get(tabId, (tab: chrome.tabs.Tab) => {
+        console.log(tab);
```
removed.

svrnm commented on browser-extension/src/instrumentation/index.ts:
```diff
@@ -0,0 +1,100 @@
+const extensionId = configTag?.dataset.extensionId;
+
+// Minimum required setup - supports only synchronous operations
+const provider = new WebTracerProvider();
```
that's a copy&paste comment from the sample, I removed it

svrnm commented on browser-extension/src/utils/EventEmitterSpanExporter.ts:
```diff
@@ -0,0 +1,46 @@
+        error,
+      });
+    }
+    this.dispatchEvent(new CustomEvent('onSpans', { detail: spans }));
```
done

svrnm commented:
I applied most of the suggestions and added some questions above. The latest commit also adds some images (and a gif) to the README.md -- a few questions from my site:

- Is the gif adding value or distracting?
- there is some "not-yet-ready" code for forwarding spans to the service worker, can this stay or should I keep it out until it is working? (I might need some help here to get the spans exported properly...)
- Any other tracers/libraries/things I should include for people to play around with?

Thanks,

vmarchaud commented:
> Is the gif adding value or distracting?

I think its worth adding gif to demo something thats visual, we sadly don't have a lot of visual things to demo generally ^^

> there is some "not-yet-ready" code for forwarding spans to the service worker, can this stay or should I keep it out until it is working? (I might need some help here to get the spans exported properly...)

I would prefer to keep it out until everything works to avoid having people open issues for it

vmarchaud reviewed (commented):
overall lgtm, not a browser expert so i'll delegate actual approve to others

svrnm commented:
Thanks everyone for looking into this:-)

> I would prefer to keep it out until everything works to avoid having people open issues for it

I removed that code for the time being and I'll reintroduce it later

> Tests are failing, but once that is resolved I think this will be ✅ from me

I fixed all of them except node:8 and browser. I can fix "browser" but I struggle with node8: webpack only supports node10+, so I am wondering if I can except the build of the extension from node8? 

Actually, I am wondering if having the extension build during the unit tests even makes any sense? There are no tests with the extension (yet) and if there would be tests, they would only need to work with the browser, since there is no value in being able to build(!) the extension in all node.js versions. Can anyone guide me here what the best way is?

obecny commented:
Last thing, I have tested this a bit and it doesn't load until you click on extension. But shouldn't this be working for each page from the very beginning. You can easily reproduce it by opening a new tab then open console then make some action on page - nothing will be exported. You have to lick on icon, and then it will be working. Not sure if that was the desired behaviour but then the icon of the extension should be disabled, and when user clicks on it the icon should be changed. Also the icon should then have some toggle functionality so you can disable this again. But tbh I would simply prefer that this will be working from start without necessity to click as this is confusing.

svrnm commented:
> Last thing, I have tested this a bit and it doesn't load until you click on extension. But shouldn't this be working for each page from the very beginning. You can easily reproduce it by opening a new tab then open console then make some action on page - nothing will be exported. You have to lick on icon, and then it will be working. Not sure if that was the desired behaviour but then the icon of the extension should be disabled, and when user clicks on it the icon should be changed. Also the icon should then have some toggle functionality so you can disable this again. But tbh I would simply prefer that this will be working from start without necessity to click as this is confusing.
> 
> Just looking at other extensions you can either disable or enable this for certain sites or simply disable for all or enable for all, maybe that is the way to go ? as a start I would just give user option to disable or enable for all so you don't have to click each time and then you can't disable too.

This behaviour is the "desired behaviour" suggested (or enforced if you want to be on the Webstore) from Google: [Substitute the activeTab permission 
](https://developer.chrome.com/docs/extensions/mv3/permission_warnings/#activeTab_permission)

Indeed you can ask for access on all sites via `optional_permissions` in manifest version 2, but I didn't figure out how to do it in manifest version 3 yet, since they explicitly introduced `host_permissions` without making it clear how to make them `optional`.

So much to the background. At the end you are 100% right. Let me do the following:

- I make "manifest version 2" (the one also supported by Chromium, Firefox, ...) the default in the code
- As suggested, I add a toggle (for mv2) that you can grant those permissions always. I have done this before, and it's not that of a big change.

There is a whole other question around, if this extension should be published in the Firefox Addons and Chrome Extension Store for easy access. To be honest, I want to keep this conversation out of this Pull Request and have it in a separate issue/discussion as a follow up.

obecny commented:
with regards to releasing this I think this is another story, we should probably have a way of somehow automate that - I guess we should have some dedicated otel account for that - something that probably @dyladan can help with to ask how this can be handled. But for now I would focus only on the extension itself.

svrnm commented:
Hey all, sorry for the delay in this (and the spam with all the checks failing yesterday...). I would need some guidance on what to do with the unit tests: as of now the extension has no tests, and from my perspective it makes no sense to verify that the extension can be build with every node.js version, especially since webpack does not support 8.x

I see two options now:
- I leave "tests" out entirely.

- I build a "browser:test" with karma as with the other web packages, although I would need some guidance here how things are interconnected.

Thanks!

dyladan commented:
Are you suggesting that the package can't be tested or shouldn't be tested?

dyladan commented:
Yes I think it is reasonable to test in browser only. You can look at other browser packages like one of the browser instrumentations to see how this is done, but it is relatively straightforward

svrnm commented:
> Yes I think it is reasonable to test in browser only. You can look at other browser packages like one of the browser instrumentations to see how this is done, but it is relatively straightforward

it was not as straightforward as assumed: There were some blockers I ran into. webpack5+karma have lead to some issues and browser extensions in karma where another complication. Anyways, I did it the otherway round and mocked the browser environment and have some tests working 🙌

vmarchaud commented on packages/opentelemetry-browser-extension-autoinjection/package.json:
```diff
@@ -0,0 +1,83 @@
+    "@opentelemetry/api": "^0.21.0"
+  },
+  "devDependencies": {
+    "@opentelemetry/api": "0.21.0",
```
i think we can directly use api 1.0 (PR to update all contrib is there: https://github.com/open-telemetry/opentelemetry-js-contrib/pull/540)

svrnm replied:
done

svrnm replied:
@vmarchaud this does not work with Node.JS 16 as it seems. I am not sure yet how to fix this since it works locally for me:

```Run npx lerna bootstrap --no-ci
lerna notice cli v3.22.1
lerna info Bootstrapping 33 packages
lerna info Installing external dependencies
lerna ERR! npm install exited 1 in '@opentelemetry/browser-extension-autoinjection'
lerna ERR! npm install stderr:
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: @opentelemetry/browser-extension-autoinjection@0.21.0
npm ERR! Found: @opentelemetry/api@1.0.0
npm ERR! node_modules/@opentelemetry/api
npm ERR!   dev @opentelemetry/api@"1.0.0" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer @opentelemetry/api@"^0.21.0" from @opentelemetry/core@0.21.0
npm ERR! node_modules/@opentelemetry/core
npm ERR!   @opentelemetry/core@"^0.21.0" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /github/home/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /github/home/.npm/_logs/2021-06-21T09_08_46_212Z-debug.log

lerna ERR! npm install exited 1 in '@opentelemetry/browser-extension-autoinjection'
lerna WARN complete Waiting for 1 child process to exit. CTRL-C to exit immediately.
Error: Process completed with exit code 1.```

vmarchaud replied:
Okay it seems we'll need to merge #540 before, since core packages peer depend on 0.12 for contrib we can't bump (i think its because of a new npm version)

vmarchaud replied:
@svrnm Should be good now #540 has been merged, could you try again ?

svrnm replied:
Sure! Let's see if that works :-)

svrnm replied:
@vmarchaud ... no, same error :-(

dyladan commented on packages/opentelemetry-browser-extension-autoinjection/package.json:
```diff
@@ -0,0 +1,83 @@
+    "@material-ui/icons": "^4.11.2",
+    "@material-ui/lab": "^4.0.0-alpha.58",
+    "@opentelemetry/context-zone": "^0.21.0",
+    "@opentelemetry/core": "^0.21.0",
+    "@opentelemetry/exporter-collector": "^0.21.0",
+    "@opentelemetry/exporter-zipkin": "^0.21.0",
+    "@opentelemetry/instrumentation": "^0.21.0",
+    "@opentelemetry/instrumentation-document-load": "^0.21.0",
+    "@opentelemetry/instrumentation-fetch": "^0.21.0",
+    "@opentelemetry/instrumentation-xml-http-request": "^0.21.0",
+    "@opentelemetry/resources": "^0.21.0",
+    "@opentelemetry/semantic-conventions": "^0.21.0",
+    "@opentelemetry/tracing": "^0.21.0",
+    "@opentelemetry/web": "^0.21.0",
```
```suggestion
    "@opentelemetry/core": "^0.22.0",
    "@opentelemetry/exporter-collector": "^0.22.0",
    "@opentelemetry/exporter-zipkin": "^0.22.0",
    "@opentelemetry/instrumentation": "^0.22.0",
    "@opentelemetry/instrumentation-document-load": "^0.21.0",
    "@opentelemetry/instrumentation-fetch": "^0.22.0",
    "@opentelemetry/instrumentation-xml-http-request": "^0.22.0",
    "@opentelemetry/resources": "^0.22.0",
    "@opentelemetry/semantic-conventions": "^0.22.0",
    "@opentelemetry/tracing": "^0.22.0",
    "@opentelemetry/web": "^0.22.0",
```

svrnm replied:
Thanks, that made it work!

obecny reviewed (approved):
I think all my concerns have been addressed, meanwhile the new version has been released so I think it can be just upgraded to the latest. Gonna approve just please address upgrade to latest before merging, gr8 job :)

svrnm commented:
> I think all my concerns have been addressed, meanwhile the new version has been released so I think it can be just upgraded to the latest. Gonna approve just please address upgrade to latest before merging, gr8 job :)

Thanks to @dyladan that's fixed already:-)

svrnm commented:
Can someone help me to understand the code coverage issue? when I click into the details I get some errors. I guess I need to increase coverage on the extension, but I would like to verify.

Flarna commented:
I guess you have to add `"codecov": "3.8.2"`  as dev dependency.

svrnm commented:
Thanks @Flarna! Now it works, @dyladan :-)

svrnm commented:
Thanks everyone for making this happen 🎉
